Staking / Ursula guide update (#1284)
nucypher/nucypher: docs/source/guides/staking_guide.rst

.. _staking-guide:

=======================
NuCypher Staking Guide
=======================

> The account which is actively doing staking operations is necessary to be a hot

Reviewer: What are we defining as "staking operations" here? Staking operations as I understand them (e.g. stake, split stake, etc.) do not require a hot wallet. If you're referring to Worker operations (e.g. re-encrypt) I'd call them something else.

Author: True, that's only stakeholder's commands. I thought to put worker commands here, but I realized that there is already a section for that. Will fix

> The account which is actively doing staking operations is necessary to be a hot

Reviewer: ```suggestion
The account which is actively doing staking operations needs to be a hot
```

wallet. But it doesn't have to be the same account as the one which receives and
sends NU tokens. The account which holds NU tokens is called StakeHolder, or

> Staker, and the account which participates in the network as Ursula node is

Reviewer: ```suggestion
Staker, and the account which participates in the network as an Ursula node is
```

called Worker.

> StakeHolder and Worker can have the same Ethereum account, but it's better to

Reviewer 2: I'd rephrase this to make it in a less positive light, saying first what is recommended (staker != worker), and only later saying that staker==worker is allowed, but not recommended.

> StakeHolder and Worker can have the same Ethereum account, but it's better to

Reviewer 3: ```suggestion
While it is possible for the StakeHolder and Worker to have the same Ethereum account, we strongly recommend keeping them separate:
```

> have them separate: StakeHolder controlled by a hardware wallet, and Worker

Reviewer 3: Associated with previous suggestion
```suggestion
 StakeHolder controlled by a hardware wallet, and Worker
```

Most stakers on the Goerli testnet will complete the following steps:

1) Install nucypher on StakeHolder node https://docs.nucypher.com/en/latest/guides/installation_guide.html

> 2) Install and run Geth, Parity or another ethereum node (can use with software or hardware Ethereum wallet).

Reviewer 3: ```suggestion
2) Install and run Geth, Parity or another ethereum node (can be used with software or hardware Ethereum wallet).
```

3) Request testnet tokens from faucet
4) Stake tokens (See Below)
5) Install another Ethereum node at the Worker instance
6) Initialize a Worker node and bond it to your Staker (``set-worker``) :ref:`ursula-config-guide`

> 7) Configure and run the Worker, and keep it online (See Below)!

Reviewer: make See Below a relative url

Interactive Method
------------------

Run an Ethereum node for stakeholder
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Assuming you have ``geth`` installed, let's run a node on Görli testnet.

Reviewer 3: We should be consistent with our spelling - we have used both `Goerli` and `Görli`.

Author: Yeah, ok, `oe` is kind of a way to spell when you cannot type `ö`. I assume, can be interchangeable?

    | ~ | Staker | Worker | # | Value    | Duration     | Enactment
    |   | ------ | ------ | - | -------- | ------------ | -----------------------------------------
    | 0 | 0xbb01 | 0xdead | 0 | 15000 NU | 41 periods . | Aug 04 12:15:16 CEST - Sep 13 12:15:16 CEST
    | 1 | 0xbb02 | 0xbeef | 1 | 15000 NU | 30 periods . | Aug 20 12:15:16 CEST - Sep 18 12:15:16 CEST
    | 2 | 0xbb03 | 0x0000 | 0 | 30000 NU | 30 periods . | Aug 09 12:15:16 CEST - Sep 9 12:15:16 CEST

If the Worker in the list is shown as ``0x0000``, it means that you haven't yet

> attach attached a Worker node to your Staker, so you still have to do it!

Reviewer 3: ```suggestion
attached a Worker node to your Staker, so you still have to do it!
```

    | ~ | Staker | Worker | # | Value    | Duration     | Enactment
    |   | ------ | ------ | - | -------- | ------------ | -----------------------------------------
    | 0 | 0xbb01 | 0xdead | 0 | 15000 NU | 41 periods . | Aug 04 12:15:16 CEST - Sep 13 12:15:16 CEST
    | 1 | 0xbb02 | 0xbeef | 1 | 15000 NU | 30 periods . | Aug 20 12:15:16 CEST - Sep 18 12:15:16 CEST
    | 2 | 0xbb03 | 0x0000 | 0 | 30000 NU | 30 periods . | Aug 09 12:15:16 CEST - Sep 9 12:15:16 CEST

> If the Worker in the list is shown as ``0x0000``, it means that you didn't yet

Reviewer: ```suggestion
If the Worker in the list is shown as ``0x0000``, it means that you haven't yet
```

> attach any node to your Staker, so you still have to do it!

Reviewer: ```suggestion
attached a Worker node to your Staker, so you still have to do it!
```

Bond an Ursula to a Staker
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> After you created an Ethereum node for your worker (see below about the worker),

Reviewer 3: ```suggestion
After you create an Ethereum node for your worker (see below about the worker),
```

> you can set the worker. The stake which doesn't yet have a worker will be

Reviewer: ```suggestion
you can set the worker. Stake which don't yet have a worker will be
```

Reviewer 3: ^ _**Stakes**_ which don't yet...

highlighted in yellow:

.. code:: bash

    (nucypher)$ nucypher stake set-worker --hw-wallet

    ======================================= Active Stakes =========================================

    | ~ | Staker | Worker | # | Value    | Duration     | Enactment
    |   | ------ | ------ | - | -------- | ------------ | -----------------------------------------
    | 0 | 0xbb01 | 0xdead | 0 | 15000 NU | 41 periods . | Aug 04 12:15:16 CEST - Sep 13 12:15:16 CEST
    | 1 | 0xbb02 | 0xbeef | 1 | 15000 NU | 30 periods . | Aug 20 12:15:16 CEST - Sep 18 12:15:16 CEST
    | 2 | 0xbb03 | 0x0000 | 0 | 30000 NU | 30 periods . | Aug 09 12:15:16 CEST - Sep 9 12:15:16 CEST

    Select Stake: 2
    Enter Worker Address: 0xbeefc4fE50f91eF73c5dD6eD89f38D55A6b1EdCA
    Worker 0xbb04c4fE50f91eF73c5dD6eD89f38D55A6b1EdCA successfully bonded to staker 0xbb03...

    OK!

Please note that the address should be in the format where checksum is encoded

> in the address. However, geth shows addresses in the lower case. You can convert

Reviewer: Is it easier for us to just accept lower case addresses rather than asking people to manually convert by copy/paste?

Author: I think so! We can do the same as we did for Felix

by copying and pasting the address to `Goerli Etherscan <https://goerli.etherscan.io/>`_, e.g.
``0x287a817426dd1ae78ea23e9918e2273b6733a43d -> 0x287A817426DD1AE78ea23e9918e2273b6733a43D``.

After this step, you're finished with the Staker, and you can proceed to :ref:`ursula-config-guide`.

Collect rewards earned by the staker
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Choose your staking address to withdraw compensation from
(``nucypher stake accounts`` will show). Then, use
``nucypher stake collect-reward`` with an option ``--staking-reward`` to collect

> inflation rewards in NU or ``--policy-reward`` to collect Ethers earned as

Reviewer: If you leave out the flag, does it collect both staking and policy rewards?

Author: Eh, right now it's a _required_ flag. Also not sure if it collects both if I put both - need to try.
I think, it's a sensible default to collect both without the flag, right?

Author: Ok, looks like if you specify both, it actually works ok

Author: :heavy_check_mark:

+----------------+----------------+--------------+

Stake 30000 NU for 90 Periods
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code:: bash

    (nucypher)$ nucypher stake init --value 30000 --duration 90 --hw-wallet
    ...

Divide stake at index 0, at 15000 NU for 30 additional Periods
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code:: bash

    (nucypher)$ nucypher stake divide --index 0 --value 15000 --duration 30 --hw-wallet
    ...

Worker configuration
------------------------

> Please go to :ref:`ursula-config-guide`.

Reviewer 3: ```suggestion
See :ref:`ursula-config-guide`.
```
